src/combinatorics/simple_graph/coloring.lean

```diff
@@ -0,0 +1,403 @@
+## Main definitions
+
+* `G.coloring α` is the type of `α`-colorings of a simple graph `G`,
+  with `α` being the set of available colors. The type is defined to
+  be homomorphisms from `G` into the complete graph on `α`.
+
+* `G.colorable n` is the proposition that `G` is `n`-colorable, which
+  is whether there exists a coloring with at most *n* colors.
+
+* `G.chromatic_number` is the minimal `n` such that `G` is
+  `n`-colorable, or `0` if it cannot be colored with finitely many
+  colors.
+
+* A *color class* is a set of vertices that share the same color.
+
+## Todo:
+
+  * Gather material from:
+    * https://github.com/leanprover-community/mathlib/blob/simple_graph_matching/src/combinatorics/simple_graph/coloring.lean
+    * https://github.com/kmill/lean-graphcoloring/blob/master/src/graph.lean
+
+  * Lowerbound for cliques
+
+  * Trees
+
+  * Planar graphs
+
+  * Chromatic polynomials
+
+  * develop API for partial colorings, likely as colorings of subgraphs (`H.coe.coloring α`)
+-/
+
+universes u v
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/--
+An `α`-coloring of a simple graph `G` is a homomorphism of `G` into the complete graph on `α`.
+This is also known as a proper coloring.
+-/
+abbreviation coloring (α : Type v) := G →g (⊤ : simple_graph α)
+
+variables {G} {α : Type v} (C : G.coloring α)
+
+lemma coloring.valid {v w : V} (h : G.adj v w) : C v ≠ C w :=
+C.map_rel h
+
+/--
+Construct a term of `simple_graph.coloring` using a function that
+assigns vertices to colors and a proof that it is as proper coloring.
+
+(Note: this is a definitionally the constructor for `simple_graph.hom`,
+but with a syntactically better proper coloring hypothesis.)
+-/
+@[pattern] def coloring.mk
+  (color : V → α)
+  (valid : ∀ {v w : V}, G.adj v w → color v ≠ color w) :
+  G.coloring α := ⟨color, @valid⟩
+
+/--
+The color class of a given color.
+-/
+def coloring.color_class_of_color (c : α) : set V := {v : V | C v = c}
+
+/-- The color class of a given vertex. -/
+def coloring.color_class_of_vertex (v : V) : set V := C.color_class_of_color (C v)
+
+/-- The set containing all color classes. -/
+def coloring.color_classes : set (set V) := {(C.color_class_of_vertex v) | v : V}
+
+lemma coloring.vertex_in_its_color_class {v : V} :
+  v ∈ {v' : V | C v' = C v} := by exact rfl
+
+lemma coloring.color_classes_is_partition :
+  setoid.is_partition C.color_classes :=
+begin
+  rw setoid.is_partition,
+  simp only [exists_unique],
+  simp [coloring.color_classes, coloring.color_class_of_vertex,
+    coloring.color_class_of_color],
+  split,
+  { intro v,
+    rw [← set.not_nonempty_iff_eq_empty, not_not],
+    use v,
+    apply coloring.vertex_in_its_color_class, },
+  { intro v,
+    split,
+    { split,
+      { use v,
+        apply coloring.vertex_in_its_color_class, },
+      { intros w hcvw,
+        rw hcvw, }, }, },
+end
+
+lemma coloring.color_classes_is_independent :
+  ∀ (s ∈ C.color_classes), is_antichain G.adj s :=
```

Comment:
It's slightly more general in this form:
```suggestion
  ∀ c, is_antichain G.adj (C.color_class c) :=
```
Also, the `c` argument can go before the lemma's colon.